src/Grids/extruded.jl

```diff
@@ -0,0 +1,120 @@
+#####
+##### Hybrid mesh
+#####
+
+abstract type HypsographyAdaption end
+
+"""
+    Flat()
+
+No surface hypsography.
+"""
+struct Flat <: HypsographyAdaption end
+
+abstract type AbstractExtrudedFiniteDifferenceGrid <: AbstractGrid end
+
+"""
+    ExtrudedFiniteDifferenceGrid(
+        horizontal_space::AbstractSpace,
+        vertical_space::FiniteDifferenceSpace,
+        hypsography::HypsographyAdaption = Flat(),
+    )
+
+Construct an `ExtrudedFiniteDifferenceGrid` from the horizontal and vertical spaces.
+"""
+mutable struct ExtrudedFiniteDifferenceGrid{
+    H <: AbstractGrid,
+    V <: FiniteDifferenceGrid,
+    A <: HypsographyAdaption,
+    GG <: Geometry.AbstractGlobalGeometry,
+    LG,
+} <: AbstractExtrudedFiniteDifferenceGrid
+    horizontal_grid::H
+    vertical_grid::V
+    hypsography::A
+    global_geometry::GG
+    center_local_geometry::LG
+    face_local_geometry::LG
+end
+
+@memoize WeakValueDict function ExtrudedFiniteDifferenceGrid(
+    horizontal_grid::Union{SpectralElementGrid1D, SpectralElementGrid2D},
+    vertical_grid::FiniteDifferenceGrid,
+    hypsography::Flat = Flat(),
+)
+    global_geometry = horizontal_grid.global_geometry
+    center_local_geometry =
+        Geometry.product_geometry.(
+            horizontal_grid.local_geometry,
+            vertical_grid.center_local_geometry,
+        )
+    face_local_geometry =
+        Geometry.product_geometry.(
+            horizontal_grid.local_geometry,
+            vertical_grid.face_local_geometry,
+        )
+
+    return ExtrudedFiniteDifferenceGrid(
+        horizontal_grid,
+        vertical_grid,
+        hypsography,
+        global_geometry,
+        center_local_geometry,
+        face_local_geometry,
+    )
+end
+
+topology(grid::ExtrudedFiniteDifferenceGrid) = topology(grid.horizontal_grid)
+
+vertical_topology(grid::ExtrudedFiniteDifferenceGrid) =
+    topology(grid.vertical_grid)
+
+local_dss_weights(grid::ExtrudedFiniteDifferenceGrid) =
+    local_dss_weights(grid.horizontal_grid)
+
+
+local_geometry_data(grid::AbstractExtrudedFiniteDifferenceGrid, ::CellCenter) =
+    grid.center_local_geometry
+local_geometry_data(grid::AbstractExtrudedFiniteDifferenceGrid, ::CellFace) =
+    grid.face_local_geometry
+global_geometry(grid::AbstractExtrudedFiniteDifferenceGrid) =
+    grid.global_geometry
```

Comment:
Is there a reason we did this for `AbstractExtrudedFiniteDifferenceGrid` over `ExtrudedFiniteDifferenceGrid`?

Reply:
needed to define it for `DeviceExtrudedFiniteDifferenceGrid` as well